cpp/modmesh/toggle/callprofiler.hpp

```diff
@@ -0,0 +1,212 @@
+ */
+
+#include <chrono>
+#include <functional>
+#include <modmesh/toggle/RadixTree.hpp>
+#include <ostream>
+#include <stack>
+#include <unordered_map>
+
+namespace modmesh
+{
+
+// The profiling result of the caller
+struct CallerProfile
+{
+    CallerProfile()
+        : total_time(0)
+    {
+    }
+
+    void start_stopwatch()
+    {
+        start_time = std::chrono::high_resolution_clock::now();
+        is_running = true;
+    }
+
+    void stop_stopwatch()
+    {
+        auto end_time = std::chrono::high_resolution_clock::now();
+        auto elapsed_time = std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start_time);
+        total_time += elapsed_time;
+        call_count++;
+    }
+
+    std::chrono::high_resolution_clock::time_point start_time;
+    std::function<void()> cancel_callback;
+    std::string caller_name;
+    std::chrono::nanoseconds total_time; /// use nanoseconds to have higher precision
+    int call_count = 0;
+    bool is_running = false;
+};
+
+class CallProfilerTest; // for gtest
+
+/// The profiler that profiles the hierarchical caller stack.
+class CallProfiler
+{
+public:
+    /// A singleton.
+    static CallProfiler & instance()
+    {
+        static CallProfiler instance;
+        return instance;
+    }
+
+    // Called when a function starts
+    void start_caller(const std::string & caller_name, std::function<void()> cancel_callback)
+    {
+        auto start_time = std::chrono::high_resolution_clock::now();
+        m_radix_tree.entry(caller_name);
+        CallerProfile & callProfile = m_radix_tree.get_current_node()->data();
+        callProfile.caller_name = caller_name;
+        callProfile.start_stopwatch();
+    }
+
+    // Called when a function ends
+    void end_caller(const std::string & caller_name)
+    {
+        CallerProfile & callProfile = m_radix_tree.get_current_node()->data();
+
+        // Update profiling information
+        callProfile.stop_stopwatch();
+
+        // Pop the caller from the call stack
+        m_radix_tree.move_current_to_parent();
+    }
+
+    /// Print the profiling information
+    void print_profiling_result(std::ostream & outstream) const
+    {
+        _print_profiling_result(*(m_radix_tree.get_current_node()), 0, outstream);
+    }
+
+    /// Result the profiler
+    void reset()
+    {
+        RadixTreeNode<CallerProfile> * newNode;
+        RadixTreeNode<CallerProfile> * currentNode = m_radix_tree.get_current_node();
+
+        while (!m_radix_tree.is_root())
+        {
+            CallerProfile & profile = m_radix_tree.get_current_node()->data();
+            if (profile.is_running)
+            {
+                profile.stop_stopwatch();
+            }
+            m_radix_tree.move_current_to_parent();
+        }
+        m_radix_tree.reset();
+    }
+
+private:
+    CallProfiler() = default;
+
+    void _print_profiling_result(const RadixTreeNode<CallerProfile> & node, const int depth, std::ostream & outstream) const
```

Comment:
Do not prefix underscore to symbols unless you are a compiler.

Reply:
And this function is long.  Please put it outside class declaration.

Reply:
done